From 083efb916af62cb0d9644f9ee1b16625ed9f997f Mon Sep 17 00:00:00 2001
From: Richard Hughes <richard@hughsie.com>
Date: Mon, 8 Nov 2010 18:04:10 +0000
Subject: [PATCH] Fix compile for PackageKit <= 0.6.11

---
 src/gpk-enum.h |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/src/gpk-enum.h b/src/gpk-enum.h
index 7ed01f9..0d98e46 100644
--- a/src/gpk-enum.h
+++ b/src/gpk-enum.h
@@ -79,6 +79,10 @@ typedef enum {
 #define PK_ERROR_ENUM_INSTALL_ROOT_INVALID				(PK_ERROR_ENUM_PROVIDE_TYPE_NOT_SUPPORTED + 1)
 #endif
 
+#if (!PK_CHECK_VERSION(0,6,11))
+#define PK_ROLE_ENUM_UPGRADE_SYSTEM					(PK_ROLE_ENUM_SIMULATE_UPDATE_PACKAGES + 1)
+#endif
+
 void		 gpk_enum_test				(gpointer	 data);
 const gchar	*gpk_role_enum_to_localised_past	(PkRoleEnum	 role)
 							 G_GNUC_CONST;
-- 
1.7.3.2

