--- gnome-packagekit-0.1.9/data/pk-application.desktop.in	Mon Mar  3 16:01:21 2008
+++ gnome-packagekit-0.1.9/data/pk-application.desktop.in	Tue Mar 11 22:29:50 2008
@@ -8,6 +8,5 @@
 Terminal=false
 Type=Application
 Categories=GNOME;GTK;System;Settings;
-NotShowIn=KDE;
 StartupNotify=true
 
--- gnome-packagekit-0.1.9/data/pk-update-viewer.desktop.in	Fri Feb 29 10:49:03 2008
+++ gnome-packagekit-0.1.9/data/pk-update-viewer.desktop.in	Tue Mar 11 22:28:21 2008
@@ -8,6 +8,5 @@
 Terminal=false
 Type=Application
 Categories=GNOME;GTK;Settings;System;PackageManager;
-NotShowIn=KDE
 StartupNotify=true
 
--- gnome-packagekit-0.1.9/data/pk-repo.desktop.in	Fri Feb 29 10:49:03 2008
+++ gnome-packagekit-0.1.9/data/pk-repo.desktop.in	Tue Mar 11 22:29:24 2008
@@ -8,6 +8,5 @@
 Terminal=false
 Type=Application
 Categories=GNOME;GTK;System;Settings;
-NotShowIn=KDE;
 StartupNotify=true
 
--- gnome-packagekit-0.1.9/data/pk-update-icon.desktop.in	Fri Feb 29 10:49:03 2008
+++ gnome-packagekit-0.1.9/data/pk-update-icon.desktop.in	Tue Mar 11 22:28:27 2008
@@ -8,5 +8,4 @@
 Terminal=false
 Type=Application
 Categories=
-NotShowIn=KDE;XFCE;
 
--- gnome-packagekit-0.1.9/data/pk-prefs.desktop.in	Fri Feb 29 10:49:04 2008
+++ gnome-packagekit-0.1.9/data/pk-prefs.desktop.in	Tue Mar 11 22:29:39 2008
@@ -8,6 +8,5 @@
 Terminal=false
 Type=Application
 Categories=Settings;X-GNOME-SystemSettings;
-NotShowIn=KDE;
 StartupNotify=true
 
