diff -up gnome-packagekit-0.1.5/src/pk-prefs.c.Werror gnome-packagekit-0.1.5/src/pk-prefs.c
--- gnome-packagekit-0.1.5/src/pk-prefs.c.Werror	2007-11-22 23:11:52.000000000 +0100
+++ gnome-packagekit-0.1.5/src/pk-prefs.c	2007-12-30 04:21:25.000000000 +0100
@@ -110,7 +110,7 @@ pk_prefs_freq_combo_changed (GtkWidget *
 {
 	gchar *value;
 	const gchar *action;
-	PkFreqEnum freq;
+	PkFreqEnum freq = PK_FREQ_ENUM_UNKNOWN;
 	GConfClient *client;
 
 	client = gconf_client_get_default ();
@@ -142,7 +142,7 @@ pk_prefs_update_combo_changed (GtkWidget
 {
 	gchar *value;
 	const gchar *action;
-	PkUpdateEnum update;
+	PkUpdateEnum update = PK_UPDATE_ENUM_UNKNOWN;
 	GConfClient *client;
 	GtkWidget *check_widget;
 	GtkWidget *notify_widget;
